extensions/amp-social-share/0.2/social-share.js

```diff
@@ -30,63 +32,196 @@ const DEFAULT_HEIGHT = 44;
+    }
+
+    // User must provide shareEndpoint if they choose a type that is not
+    // pre-configured
+    const typeConfig = getSocialConfig(type) || dict();
+    const baseEndpoint = typeConfig['shareEndpoint'] || shareEndpoint;
+    if (baseEndpoint === undefined) {
+      throwError_(
+        `A shareEndpoint is required if not using a pre-configured type. ${name}`
+      );
+    }
+
+    // Throw warning if type is obsolete
+    const obsoleteType = typeConfig['obsolete'];
+    if (obsoleteType) {
+      throwWarning_(`Skipping obsolete share button ${type}. ${name}`);
     }
+
+    // Verify width and height are valid integers
+    let checkedWidth = width || DEFAULT_WIDTH;
+    let checkedHeight = height || DEFAULT_HEIGHT;
+    if (width && !Number.isInteger(width)) {
+      throwWarning_(
+        `The width property should be an Integer, defaulting to ${DEFAULT_WIDTH}. ${name}`
+      );
+      checkedWidth = DEFAULT_WIDTH;
+    }
+    if (height && !Number.isInteger(height)) {
+      throwWarning_(
+        `The height property should be an Integer, defaulting to ${DEFAULT_HEIGHT}. ${name}`
+      );
+      checkedHeight = DEFAULT_HEIGHT;
+    }
+    return {
+      typeConfig,
+      baseEndpoint,
+      checkedWidth,
+      checkedHeight,
+      obsoleteType,
+    };
   }
 
-  /** @private */
-  function handleActivation() {
-    const href = props['href'];
-    const target = props['target'] || '_blank';
-    if (!href) {
-      throw new Error('Clicked before href is set.');
+  /**
+   * @private
+   * @param {!JsonObject} typeConfig
+   * @param {?string} baseEndpoint
+   * @param {!JsonObject} props
+   * @return {?string}
+   */
+  function createEndpoint_(typeConfig, baseEndpoint, props) {
+    const {params, bindings} = props;
+    const combinedParams = dict();
+    Object.assign(combinedParams, typeConfig['defaultParams'], params);
+    const endpointWithParams = addParamsToUrl(baseEndpoint, combinedParams);
+
+    const combinedBindings = dict();
+    const bindingVars = typeConfig['bindings'];
+    if (bindingVars) {
+      bindingVars.forEach((name) => {
+        combinedBindings[name.toUpperCase()] = params[name] ? params[name] : '';
+      });
     }
-    if (startsWith(href, 'navigator-share:')) {
-      if (!navigator.share) {
-        throw new Error('no navigator.share');
-      }
-      const dataStr = href.substr(href.indexOf('?'));
-      const data = parseQueryString(dataStr);
-      navigator.share(data).catch(() => {
-        // TODO(alanorozco): Warn here somehow.
-        // warn(TAG, e.message, dataStr);
+    if (bindings) {
+      Object.keys(bindings).forEach((name) => {
+        combinedBindings[name.toUpperCase()] = bindings[name]
+          ? bindings[name]
+          : '';
       });
+    }
+    const finalEndpoint = Object.keys(combinedBindings).reduce(
+      (endpoint, binding) =>
+        endpoint.replace(new RegExp(binding, 'g'), combinedBindings[binding]),
```

Comment:
So, if we are changing anything here, I'd rather we make the config as convenient for us to use as possible. IMHO, why not just use functions in the config itself to expand bindings? It's really there with one purpose only: making our component nice and small. This code makes an impression that we have to make component more complicated just to support a specific and inconvenient config format.

Reply:
Sure, I'm open to updating the config file.  Maybe we should discuss further in real time?

Reply:
I described it [here](https://github.com/ampproject/amphtml/pull/28137#discussion_r419730258). f2f discussion is great. I'll close this comment so that we can dedup the threads.